### DISCUSSION TITLE
Looking for assistance on replicating existing usage of GitPython code (to replace with dulwich)

### DISCUSSION THREAD

**Discussion by EdgyEdgemond** [jelmer/dulwich#1345] [General]
I've been digging round the docs and the code and have found some of the answers I am looking for, and others I can't get the full functionality I am after, so hoping theres a different way to achieve the same outcome.

Im including all the commands i'm looking to replicate for completeness. These are all executed in code via subprocess, so no need for direct cli replication, just a pointer to how this would be achieved in dulwich, if it can be achieved.

```
git rev-list HEAD..origin/main
git rev-list origin/main..HEAD
git status -s                                # viable solution found
git tag -l *X.Y.Z                            # viable solution found
git log --format=%h:%H:%B -z [tag]..HEAD
git add --update [*path]                     # viable solution found
git reset HEAD~1 --hard
git commit -m [message]                      # viable solution found
git tag vX.Y.Z                               # viable solution found
```

dulwich rev-list wants a commit id, can't see a way to check for the diff between remote and local (happy to use plumbing over porcelain if the functionality can be replicated, just need pointing to where to look)

porcelain.status can replicate what I need from `git status -s` but is 8 times slower than subprocess call to equivalent git command.

porcelain.tag does not appear to sort listing (or filtering) edit: Found tag_list.

porcelain.log does not appear to support formatting, or changing the separator edit: might not need either of those as have raw commits, but can't work out how to get logs between a range.

have not looked into add/reset/commit/tag yet, but can see plumbing for those, so don't expect those to be as much of an issue.

**Comment by EdgyEdgemond** [comment]
My adaptation of porcelain.status to try to reduce time (6x `git status -s`) 

```
            index = self.repo.open_index()
            try:
                tree_id = self.repo[b"HEAD"].tree
            except KeyError:
                tree_id = None

            path = self.repo.path
            staged = next(index.changes_from_tree(self.repo.object_store, tree_id), None)
            unstaged = next(porcelain.get_unstaged_changes(index, path), None)
            untracked = next(porcelain.get_untracked_paths(path, path, index, path, untracked_files="all"), None)
            dirty = {staged, unstaged, untracked} != {None}
```

**Reply by EdgyEdgemond** [reply]
attempting `git reset HEAD~1 --hard`, using `porcelain.reset(...)` or equivalent underlying code, looks to revert the changes and leave them unstaged but HEAD remains the same.

```
            c = self.repo[self.repo.head()]    # <Commit b'731cbd12e1222c4e81834c5a0e1b3fc96fb9e9f2'>
            t = self.repo[c.tree]              # <Tree b'f9e35f9776c72d0fddcfe64bcf3e3c1199456f8c'>
            p = self.repo[c.parents[0]]        # <Commit b'1381fab18ce967662b647234eabf72f3d8a3b3d3'>
            t = self.repo[p.tree]              # <Tree b'ef85ac1334f9a366f819795d7a79bfe34a5482b7'>
            self.repo.reset_index(t.id)
            c = self.repo[self.repo.head()]    # <Commit b'731cbd12e1222c4e81834c5a0e1b3fc96fb9e9f2'>
```
```
$ git log
commit 731cbd12e1222c4e81834c5a0e1b3fc96fb9e9f2 (HEAD)

$ git status
...
Changes to be committed:
  (use "git restore --staged <file>..." to unstage)
	modified:   README.md
```

Getting dulwich HEAD either side of `git reset HEAD~1 --hard` using GitPython gets the expected changed heads.

**Comment by jelmer** [comment]
So dulwich' porcelain module is still fairly limited in terms of what it supports. Here's some ideas for the commands you're looking at:

> git rev-list HEAD..origin/main
> git rev-list origin/main..HEAD

dulwich.porcelain.rev_list() should ideally take a range but today doesn't.

You can use Repo.get_walker() and pass in include=['HEAD'], exclude=['refs/heads/origin/main']

> git log --format=%h:%H:%B -z [tag]..HEAD

This isn't supported at the moment (see dulwich.porcelain.log) although you could copy that function and modify it to print commits with that format. 

> git reset HEAD~1 --hard

dulwich.porcelain.reset(mode="hard") only supports resetting the index at the moment.
